views/Data.js

```diff
@@ -251,39 +251,79 @@ export function useData({
   formatOut = null,
   validate = null,
   validateRequired = false,
-  viewPath = null,
 } = {}) {
-  let [data, dispatch, onSubmit, originalValue] = useContext(
-    DataContexts[context]
+  let { value } = useDataFormat({ context, path, formatIn })
+  let { originalValue } = useDataOriginalValue({
+    context,
+  })
+  let { onChange } = useDataOnChange({ context, path, formatOut })
+  let { onSubmit } = useDataOnSubmit({ context })
+  let { isSubmitting } = useDataIsSubmitting({ context })
+  let {
+    isValid,
+    isValidInitial,
+    isInvalid,
+    isInvalidInitial,
+  } = useDataValidate({ context, path, validate, validateRequired })
+
+  return {
+    value,
+    originalValue,
+    onChange,
+    onSubmit,
+    isSubmitting,
+    isValid,
+    isValidInitial,
+    isInvalid,
+    isInvalidInitial,
+  }
+}
+
+export function useDataValue({ context = 'default', path = null } = {}) {
+  let [data] = useContext(DataContexts[context])
+  let value = useMemo(() => (path ? get(data.value, path) : data.value), [
+    path,
+    data.value,
+  ])
+  return { value }
+}
+
+export function useDataOriginalValue({ context = 'default' } = {}) {
+  let [, , , originalValue] = useContext(
+    DataContexts[context] || DataContexts.default
```

Comment:
if context = `default`, shouldn't this:
```DataContexts[context] || DataContexts.default```
be:
```DataContexts[context]```
?

Reply:
Ah right, yeah, well, that probably makes sense for dev mode then if anything so it doesn't trip.
In any case, what about abstracting the context call to a hook of its own that deals with this?
```
function useDataContext(context) {
  return useContext(DataContexts[context] || DataContexts.default)
}
```
That way we can call it from every other bit without having to worry about it. In prod we can have:
```
function useDataContext(context) {
    return useContext(DataContexts[context])
}

```

And in the Tools file we could have:
```
function useDataContext(name) {
  let context = DataContexts[name]
  if (!context) {
     context = DataContexts.default
     console.error({ type: 'views/data/missing-context', context: name })
  }
  return useContext()
}
```

That way we could remove all the `default` references from the production files. What do you think?

Reply:
I like this approach, thanks. will make that change